### DISCUSSION TITLE
Supporting push notifications on Android

### DISCUSSION THREAD

**Discussion by gzzo** [react-native-community/discussions-and-proposals#360] [Ideas]
### Introduction

PushNotificationIOS works perfectly on iOS.  React Native should support push notifications on Android as well.

### The Core of It

It looks like the go-to library for Android support has been [react-native-push-notification](https://github.com/zo0r/react-native-push-notification). However, it looks like that repository isn't maintained anymore and may have been abandoned.

I think it's reasonable to expect developers to want push notification support on both platforms mostly out-of-the-box without having to resort to third parties. There's several third parties with freemium react-native support (Firebase, OneSignal, Pusher).

### Discussion points

Options:
 1. Adopt react-native-push-notification into react-native-community
 2. Create a PushNotificationAndroid API
 3. Add Android support into PushNotificationIOS and rename the API to PushNotification

**Comment by cpojer** [comment]
We are currently extracting this functionality from React Native into https://github.com/react-native-community/react-native-push-notification-ios

Would you be able to lead the effort to bring the Android version into the react-native-community and possibly building one library to handle push notifications for both platforms?

**Comment by sibelius** [comment]
you should check react-native-firebase for this

**Comment by gzzo** [comment]
@sibelius I mentioned that in the issue description. I don't think RN should rely on third parties with freemium models for Android push notifications.

@cpojer So is the preferred option from my description the third one, to have a unified API, `PushNotification`?

**Comment by sibelius** [comment]
firebase is de facto solution for android push

GCM is deprecated, everything now is FCM, you can use it for free forever

Google does not have any other option

**Comment by gzzo** [comment]
Thanks, I didn't know that. Looks like GCM will be deprecated 4/11/2019.

That means that the unified PushNotification API will only need to support FCM.

**Comment by rhdeck** [comment]
I contributed the FCM support PR for react-native-push-notification and I
use it in production.    Glad to belp a RNC initiative for that support.

**Comment by rafaellincoln** [comment]
@gzzo Feel free to open a PR on [push-notification-ios](https://github.com/react-native-community/react-native-push-notification-ios)

**Comment by gourav3309** [comment]
Hey, I am working on a React Native mobile app, and I want to integrate Firebase into my code. I am able to get the FCM token on an Android device, but I am unable to get the FCM token on an iOS device. Can you please help me? Here are my Podfile, AppDelegate.mm file, and AppDelegate.h file. Please check and provide me with a better solution as soon as possible. : 


***********AppDelegate.mm************* :: 
#import "AppDelegate.h"
#import <Firebase.h>
#import "RNFBMessagingModule.h"
#import <React/RCTBundleURLProvider.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  self.moduleName = @"crc_app";
  // You can add your custom initial props in the dictionary below.
  // They will be passed down to the ViewController used by React Native.
  self.initialProps = @{};

  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
  return [self bundleURL];
}

- (NSURL *)bundleURL
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index"];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

@end  

**************podfile : **************
# Resolve react_native_pods.rb with node to allow for hoisting
require Pod::Executable.execute_command('node', ['-p',
  'require.resolve(
    "react-native/scripts/react_native_pods.rb",
    {paths: [process.argv[1]]},
  )', __dir__]).strip

platform :ios, min_ios_version_supported
prepare_react_native_project!

$RNFirebaseAsStaticFramework = true

  pod 'Firebase', :modular_headers => true
  pod 'FirebaseCoreInternal', :modular_headers => true
  pod 'GoogleUtilities', :modular_headers => true
  pod 'FirebaseCore', :modular_headers => true
linkage = ENV['USE_FRAMEWORKS']
if linkage != nil
  Pod::UI.puts "Configuring Pod with #{linkage}ally linked Frameworks".green
  use_frameworks! :linkage => linkage.to_sym
  use_frameworks! :linkage => :static

end

target 'crc_app' do
  config = use_native_modules!

  use_react_native!(
    :path => config[:reactNativePath],
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )

  target 'crc_appTests' do
    inherit! :complete
    # Pods for testing
  end

  post_install do |installer|
    # https://github.com/facebook/react-native/blob/main/packages/react-native/scripts/react_native_pods.rb#L197-L202
    react_native_post_install(
      installer,
      config[:reactNativePath],
      :mac_catalyst_enabled => false,
      # :ccache_enabled => true
    )
  end
end 
**************AppDelegate.h :**************** 
#import <RCTAppDelegate.h>
#import <UIKit/UIKit.h>

@interface AppDelegate : RCTAppDelegate

@end
